starboard/shared/starboard/player/filter/audio_discard_duration_tracker.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright 2023 The Cobalt Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "starboard/shared/starboard/player/filter/audio_discard_duration_tracker.h"
+
+#include <algorithm>
+
+#include "starboard/common/log.h"
+
+namespace starboard {
+namespace shared {
+namespace starboard {
+namespace player {
+namespace filter {
+
+void AudioDiscardDurationTracker::CacheMultipleDiscardDurations(
+    const InputBuffers& input_buffers,
+    SbTime buffer_duration) {
+  ScopedLock lock(mutex_);
+  for (const auto& input_buffer : input_buffers) {
+    CacheDiscardDuration(input_buffer, buffer_duration);
+  }
+}
+
+SbTime AudioDiscardDurationTracker::AdjustTimeForTotalDiscardDuration(
+    SbTime timestamp) {
+  ScopedLock lock(mutex_);
+  SB_LOG_IF(WARNING, last_received_timestamp_ > timestamp)
+      << "Last received timestamp " << last_received_timestamp_
+      << " is greater than timestamp " << timestamp
+      << ". AudioDiscardDurationTracker::AdjustTimeForTotalDiscardDuration() "
+         "requires monotonically increasing timestamps.";
+  last_received_timestamp_ = timestamp;
+
```

Comment:
`discard_start_timestamp` is based on adjusted timestamp, but the input `timestamp` is not adjusted. It should be adjusted before comparing to the `discard_start_timestamp`. And the return value is `discard_start_timestamp - total_discard_duration_`, that adjusts the timestamp on an adjusted timestamp.

For example, if there're two inputs and both have 20ms audio and discarded duration from front of 5ms, the total played time would be 40ms(wall time), and `discard_infos_` would be {0ms, 5ms}, {15ms, 5ms}. During 0-15ms (wall time), the returned value works as expected. During 15-20ms(wall time), the returned value would be 10ms, which is not right. It's expected to freeze the timestamp at 20-25ms(wall time).